File: machine_learning/API/app/data_utils.py
import numpy as np

def haversine(lat1, lon1, lat2, lon2):
    """Menghitung jarak antara dua titik koordinat menggunakan rumus Haversine."""
    lat1, lon1, lat2, lon2 = map(np.radians, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = np.sin(dlat / 2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2)**2
    c = 2 * np.arcsin(np.sqrt(a))
    r = 6371  # Radius bumi dalam kilometer
    return c * r

def preprocess_data(data_penerima, data_makanan):
    """Mengolah data penerima dan makanan menjadi fitur yang ter-encode."""
    processed_data = []
    id_penerima_list = []

    for penerima in data_penerima:
        id_penerima_list.append(penerima['id_penerima'])
        for makanan in data_makanan:
            feature = [
                makanan['jumlah_disumbangkan'],
                1 if makanan['is_halal_makanan'] else 0,
                1 if makanan['is_for_child_makanan'] else 0,
                1 if makanan['is_for_elderly_makanan'] else 0,
                1 if makanan['is_alergan_makanan'] else 0,
                penerima['jumlah_dibutuhkan'],
                penerima['frekuensi_menerima'],
                1 if penerima['is_halal_receiver'] else 0,
                1 if penerima['is_for_child_receiver'] else 0,
                1 if penerima['is_for_elderly_receiver'] else 0,
                1 if penerima['is_alergan_free'] else 0,
                1 if makanan['makanan_disumbangkan'] == 'makanan' else 0,
                1 if makanan['makanan_disumbangkan'] == 'makanan_minuman' else 0,
                1 if makanan['makanan_disumbangkan'] == 'minuman' else 0,
                1 if makanan['kondisi_makanan'] == 'hampir_kadaluarsa' else 0,
                1 if makanan['kondisi_makanan'] == 'layak_konsumsi' else 0,
                1 if makanan['kondisi_makanan'] == 'tidak_layak_konsumsi' else 0,
                1 if penerima['makanan_dibutuhkan'] == 'makanan' else 0,
                1 if penerima['makanan_dibutuhkan'] == 'makanan_minuman' else 0,
                1 if penerima['makanan_dibutuhkan'] == 'minuman' else 0,
                1 if penerima['kondisi_makanan_diterima'] == 'hampir_kadaluarsa' else 0,
                1 if penerima['kondisi_makanan_diterima'] == 'layak_konsumsi' else 0,
                1 if penerima['kondisi_makanan_diterima'] == 'layak_konsumsi_hampir_kadaluarsa' else 0,
                1 if penerima['kondisi_makanan_diterima'] == 'tidak_layak_konsumsi' else 0,
                1 if penerima['status_penerima'] == 'mendesak' else 0,
                1 if penerima['status_penerima'] == 'normal' else 0,
                1 if penerima['status_penerima'] == 'tidak_mendesak' else 0,
                haversine(
                    penerima['lokasi_lat_penerima'], penerima['lokasi_lon_penerima'],
                    makanan['lokasi_lat_makanan'], makanan['lokasi_lon_makanan']
                )
            ]
            processed_data.append(feature)
    return id_penerima_list, processed_data

File: cloud/data_utils.js
const haversine = (lat1, lon1, lat2, lon2) => {
  const toRadians = (degrees) => (degrees * Math.PI) / 180;
  const R = 6371; // Radius bumi dalam kilometer
  const dLat = toRadians(lat2 - lat1);
  const dLon = toRadians(lon2 - lon1);

  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRadians(lat1)) * Math.cos(toRadians(lat2)) * Math.sin(dLon / 2) ** 2;

  return 2 * R * Math.asin(Math.sqrt(a));
};

const preprocessData = (dataPenerima, dataMakanan) => {
  const processedData = [];
  const idPenerimaList = [];

  dataPenerima.forEach((penerima) => {
    idPenerimaList.push(penerima.id_penerima);
    dataMakanan.forEach((makanan) => {
      const feature = [
        makanan.jumlah_disumbangkan,
        makanan.is_halal_makanan ? 1 : 0,
        makanan.is_for_child_makanan ? 1 : 0,
        makanan.is_for_elderly_makanan ? 1 : 0,
        makanan.is_alergan_makanan ? 1 : 0,
        penerima.jumlah_dibutuhkan,
        penerima.frekuensi_penerima,
        penerima.is_halal_receiver ? 1 : 0,
        penerima.is_for_child_receiver ? 1 : 0,
        penerima.is_for_elderly_receiver ? 1 : 0,
        penerima.is_alergan_free ? 1 : 0,
        makanan.makanan_disumbangkan === "makanan" ? 1 : 0,
        makanan.makanan_disumbangkan === "makanan_minuman" ? 1 : 0,
        makanan.makanan_disumbangkan === "minuman" ? 1 : 0,
        makanan.kondisi_makanan === "hampir_kadaluarsa" ? 1 : 0,
        makanan.kondisi_makanan === "layak_konsumsi" ? 1 : 0,
        makanan.kondisi_makanan === "tidak_layak_konsumsi" ? 1 : 0,
        penerima.makanan_dibutuhkan === "makanan" ? 1 : 0,
        penerima.makanan_dibutuhkan === "makanan_minuman" ? 1 : 0,
        penerima.makanan_dibutuhkan === "minuman" ? 1 : 0,
        penerima.kondisi_makanan_diterima === "hampir_kadaluarsa" ? 1 : 0,
        penerima.kondisi_makanan_diterima === "layak_konsumsi" ? 1 : 0,
        penerima.kondisi_makanan_diterima === "tidak_layak_konsumsi" ? 1 : 0,
        penerima.status_penerima === "mendesak" ? 1 : 0,
        penerima.status_penerima === "normal" ? 1 : 0,
        penerima.status_penerima === "tidak_mendesak" ? 1 : 0,
        haversine(
          penerima.lokasi_lat_penerima,
          penerima.lokasi_lon_penerima,
          makanan.lokasi_lat_makanan,
          makanan.lokasi_lon_makanan
        ),
      ];
      processedData.push(feature);
    });
  });

  return [idPenerimaList, processedData];
};

module.exports = { preprocessData };
